nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -975,267 +991,170 @@ ngx_http_modsecurity_init_process(ngx_cycle_t *cycle)
-            buf->last_buf = 0;
-            e = apr_bucket_eos_create(bb->bucket_alloc);
-            if (e == NULL) {
-                return ngx_http_filter_finalize_request(r, 
-                         &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
-            }
-            APR_BRIGADE_INSERT_TAIL(bb, e);
-            break;
+        if (rc >= NGX_HTTP_SPECIAL_RESPONSE) {
+            return rc;
         }
 
-        buf->pos = buf->last;
-    }
-
-    if (!last_buf) {
-        return NGX_AGAIN;
-    }
-
-    /* last buf has been saved */
-    ctx->complete = 1;
-    modsecSetResponseBrigade(ctx->req, ctx->brigade);
-
-    if (ngx_http_modsecurity_load_headers_in(r) != NGX_OK
-            || ngx_http_modsecurity_load_headers_out(r) != NGX_OK) {
-
-        return ngx_http_filter_finalize_request(r, 
-                 &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
+        return NGX_DONE;
     }
 
-    rc = ngx_http_modsecurity_status(r, modsecProcessResponse(ctx->req));
-
-    if (rc != NGX_DECLINED) {
-        return ngx_http_filter_finalize_request(r, &ngx_http_modsecurity, rc);
-    }
+    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "modSecurity: handler");
 
-    apr_brigade_length(ctx->brigade, 0, &content_length);
+    // Get module request context or create if not yet created
+    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
+    if (ctx == NULL) {
+        ctx = ngx_http_modsecurity_create_ctx(r);
+        if (ctx == NULL) {
+            return NGX_HTTP_INTERNAL_SERVER_ERROR;
+        }
 
-    rc = move_brigade_to_chain(ctx->brigade, &out, r->pool);
-    if (rc == NGX_ERROR) {
-        return ngx_http_filter_finalize_request(r, 
-                 &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
+        ngx_http_set_ctx(r, ctx, ngx_http_modsecurity);
     }
 
-    if (ngx_http_modsecurity_save_headers_in(r) != NGX_OK
-            ||ngx_http_modsecurity_save_headers_out(r) != NGX_OK) {
-
-        return ngx_http_filter_finalize_request(r, 
-                 &ngx_http_modsecurity, NGX_HTTP_INTERNAL_SERVER_ERROR);
+    // Sometimes Nginx calls ngx_http_modsecurity_handler multiple times for the same request, after a worker thread has already been started. This is to guard against it.
+    if (ctx->thread_running) {
+        return NGX_DONE;
     }
 
-    if (r->headers_out.content_length_n != -1) {
+    if (ctx->status_code != STATUS_CODE_NOT_SET) {
+        if (ctx->status_code > 0) {
+            return ctx->status_code;
+        }
 
-        r->headers_out.content_length_n = content_length;
-        r->headers_out.content_length = NULL; /* header filter will set this */
+        // Request must be routed to the next handler
+        return NGX_DECLINED;
     }
 
-    r->header_sent = 0;
-    rc = ngx_http_next_header_filter(r);
-
-    if (rc == NGX_ERROR || rc > NGX_OK) {
-        return ngx_http_filter_finalize_request(r, &ngx_http_modsecurity, rc);
+    ctx->thread_running = 1;
```

Comment:
Should this be moved after the if after it? Should we confirm that the thread creation was successful?

Reply:
We already return NGX_ERROR if thread creation was not successful.
Whether thread_running = 1 should be before or after the call to _task_offload I don't think really matters. It's my internal flag only used right here (the if a few lines above). I initially had it after, but then after Yanshu's suggestion, I moved it in front, because I was thinking maybe a new event comes in during the time we spend posting the task... It's not the most elegant workaround. Would be great to understand why these extra events that trigger _handler again can even happen, but I haven't been able to figure this out.